circuits/circom/prover/wasm/witness_calculator.js

```diff
@@ -0,0 +1,330 @@
+module.exports = async function builder(code, options) {
+  options = options || {}
+
+  let wasmModule
+  try {
+    wasmModule = await WebAssembly.compile(code)
+  } catch (err) {
+    console.log(err)
+    console.log(
+      '\nTry to run circom --c in order to generate c++ code instead\n',
+    )
+    throw new Error(err)
+  }
+
+  let wc
+
+  let errStr = ''
+  let msgStr = ''
+
+  const instance = await WebAssembly.instantiate(wasmModule, {
+    runtime: {
+      exceptionHandler: function (code) {
+        let err
+        if (code == 1) {
+          err = 'Signal not found.\n'
+        } else if (code == 2) {
+          err = 'Too many signals set.\n'
+        } else if (code == 3) {
+          err = 'Signal already set.\n'
+        } else if (code == 4) {
+          err = 'Assert Failed.\n'
+        } else if (code == 5) {
+          err = 'Not enough memory.\n'
+        } else if (code == 6) {
+          err = 'Input signal array access exceeds the size.\n'
+        } else {
+          err = 'Unknown error.\n'
+        }
+        throw new Error(err + errStr)
+      },
+      printErrorMessage: function () {
+        errStr += getMessage() + '\n'
+        // console.error(getMessage());
+      },
+      writeBufferMessage: function () {
+        const msg = getMessage()
+        // Any calls to `log()` will always end with a `\n`, so that's when we print and reset
+        if (msg === '\n') {
+          console.log(msgStr)
+          msgStr = ''
+        } else {
+          // If we've buffered other content, put a space in between the items
+          if (msgStr !== '') {
+            msgStr += ' '
+          }
+          // Then append the message to the message we are creating
+          msgStr += msg
+        }
+      },
+      showSharedRWMemory: function () {
+        printSharedRWMemory()
+      },
+    },
```

Comment:
for better readability, what about extracting these helpers functions in separate file(s)?
`exceptionHandler`, `printErrorMessage`, `writeBufferMessage`, `showSharedMemory`

Reply:
autogenerated, as above https://github.com/privacy-scaling-explorations/e2e-zk-ecdsa/pull/83#discussion_r1133407276